File: Aran/Panda/Arrival/trunk/Modules/ListViewColumnSorter.vb
Imports System.Windows.Forms.ListViewItem

Public Class ListViewColumnSorter
	Implements IComparer

	Dim ObjectCompare As CaseInsensitiveComparer

	Public Sub New()
		_columnToSort = 0
		_sortOrder = SortOrder.None
		ObjectCompare = New CaseInsensitiveComparer()
	End Sub

	Public Function Compare(x As Object, y As Object) As Integer Implements IComparer.Compare
		If _sortOrder = SortOrder.None Then Return 0

		Try
			Dim listviewX As ListViewItem = CType(x, ListViewItem)
			Dim listviewY As ListViewItem = CType(y, ListViewItem)

			If (_columnToSort > listviewX.SubItems.Count Or _columnToSort > listviewY.SubItems.Count) Then Return 0

			Dim compareResult As Integer
			Dim column1 As ListViewSubItem = listviewX.SubItems(_columnToSort)
			Dim column2 As ListViewSubItem = listviewY.SubItems(_columnToSort)

			If (IsNumeric(column1.Text) And IsNumeric(column2.Text)) Then
				Dim val1 As Double = Convert.ToDouble(column1.Text)
				Dim val2 As Double = Convert.ToDouble(column2.Text)
				compareResult = val1.CompareTo(val2)
			Else
				compareResult = ObjectCompare.Compare(listviewX.SubItems(_columnToSort).Text, listviewY.SubItems(_columnToSort).Text)
			End If

			If (_sortOrder = SortOrder.Ascending) Then Return compareResult
			Return -compareResult
		Catch
			Return 0
		End Try
	End Function

	Private _sortOrder As SortOrder
	Public Property Order() As SortOrder
		Get
			Return _sortOrder
		End Get

		Set(ByVal value As SortOrder)
			_sortOrder = value
		End Set
	End Property

	Private _columnToSort As Integer
	Public Property ColumntToSort() As Integer
		Get
			Return _columnToSort
		End Get

		Set(ByVal value As Integer)
			_columnToSort = value
		End Set
	End Property
End Class

File: Aran/Panda/Departure/trunk/Modules/ListViewColumnSorter.cs
using System.Collections;
using System.Windows.Forms;

namespace Aran.PANDA.Departure
{
	//public struct ListViewsTag
	//{
	//	public ObstacleContainer obstacles;
	//	public int SortField;
	//}

	public class ListViewColumnSorter : IComparer
	{
		public ListViewColumnSorter()
		{
			ColumnToSort = 0;
			Order = SortOrder.Descending;
			_objectComparer = new CaseInsensitiveComparer();
		}

		public int Compare(object x, object y)
		{
			if (Order == SortOrder.None)
				return 0;

			try
			{
				ListViewItem listviewX = (ListViewItem)x;
				ListViewItem listviewY = (ListViewItem)y;

				if (ColumnToSort > listviewX.SubItems.Count || ColumnToSort > listviewY.SubItems.Count)
					return 0;

				string textX = listviewX.SubItems[ColumnToSort].Text;
				string textY = listviewY.SubItems[ColumnToSort].Text;
				int compareResult;

				if (double.TryParse(textX, out double valX) && double.TryParse(textY, out double valY))
					compareResult = valX.CompareTo(valY);
				else
					compareResult = _objectComparer.Compare(textX, textY);

				if (Order == SortOrder.Ascending)
					return compareResult;

				return -compareResult;
			}
			catch
			{
				return 0;
			}
		}

		public void SortListView(int columnIndex, ListView listview)
		{
			ColumnToSort = System.Math.Abs((int)listview.Tag) - 1;
			listview.Tag = columnIndex + 1;

			if (columnIndex != ColumnToSort)
			{
				listview.Columns[ColumnToSort].ImageIndex = 2;
				ColumnToSort = columnIndex;

				Order = SortOrder.Descending;
				listview.Columns[ColumnToSort].ImageIndex = 1;
			}
			else if (Order == SortOrder.Ascending)
			{
				Order = SortOrder.Descending;
				listview.Columns[ColumnToSort].ImageIndex = 1;
			}
			else
			{
				Order = SortOrder.Ascending;
				listview.Columns[ColumnToSort].ImageIndex = 0;
			}

			listview.Sort();
		}

		private CaseInsensitiveComparer _objectComparer;
		public SortOrder Order { get; set; }
		public int ColumnToSort { get; set; }
	}
}
